### PR TITLE
feat(ios): detect if app is running on a macOS desktop with Apple Silicon

### PR DESCRIPTION
**Pull request by SailingSteve**
# Platforms affected
iOS 14 and up

### Motivation and Context
When running on iOS14, on a Mac that has an Apple Silicon processor (ARM64 instead of Intel), this change allows you to determine this condition using cordova-plugin-device.

This minor change exposes the iOS `isiOSAppOnMac` variable, which is the definitive way to detect that the Cordova iOS app is running on the Mac desktop on a non-Intel processor.

On Cordova apps running on the Apple Silicon desktop, the hardware menu (wifi, clock, cell strength, battery) goes away, and other container changes need to be compensated for, so access to this variable is essential.  This plugin seems like the best place to expose this variable.

### Description
in the iOS NSProcessInfo under iOS 14+, a new system variable has been defined, this change exposes that variable.

    [NSProcessInfo processInfo].isiOSAppOnMac

### Testing
This change works on iOS 14 on macOS Big Sur Version 11.0 Beta 8 (20A5374i) running on an Apple A12Z processor.
The change causes no problem (returns false) on macOS Catalina Version 10.15.6 running on an Intel processor inside an iOS simulator (SE 2nd Gen, 8 Plus 13.3, 11 Pro 14.0, 11 13.3, iPadPro 9.7" 14.0, iPadPro 12" 4th gen 14.0), and on a physical 
iPhone Xs Max running iOS 13.7.
Also no problem on an Android simulator for a Galaxy Tab S5e, and a Galaxy S9+ at API 30, and a Moto G5 plus (Android 8.1.10) physical phone.

### Checklist

- [ ] I've run the tests to see all new and existing tests pass
- [ ] I added automated test coverage as appropriate for this change
- [x] Commit is prefixed with `(platform)` if this change only applies to one platform (e.g. `(android)`)
- [ ] If this Pull Request resolves an issue, I linked to the issue in the text above (and used the correct [keyword to close issues using keywords](https://help.github.com/articles/closing-issues-using-keywords/))
- [x] I've updated the documentation if necessary

### CODE REVIEW

**Review by timbru31 — APPROVED** [review]
LGTM, thanks!

**Review by NiklasMerz — APPROVED** [review]
Good to me, too. This property is strangely named on Apples side IMHO. It is `false` for iOS apps running on Intel Macs using Catalyst: https://github.com/apache/cordova-plugin-device/pull/133

Not that clear on first sight but makes sense if you need to know that.

**Comment by NiklasMerz** [comment]
Thinking about this again I am confused again. Why would I want to know if the app runs on Apple ARM chips? It should not make a difference if it runs on Mac using Catalyst or natively with ARM, shouldn't it? I may need to know if it runs on desktop or mobile to adjust my styling etc?

Shouldn't the platform return something different if it runs on macOS? Technically it still runs the iOS platform but it's runnig on a Mac using Catalyst or Big Sur.

Other opinions?

**Comment by erisu** [comment]
I personally don't see what is the need of this new property. I feel that using both the `device.model` and `device.platform` should be enough to identify what is running on what device. 

To recap from the documents printed in the README of this repo,

`device.platform` should return either `iOS` or `Mac OS X`. But since we are talking about building an iOS app with the Cordova iOS platform, `iOS` will always be returned. This value is hardcoded.

`device.model` should return one of the following two cases:

```
// iOS:     for the iPad Mini, returns iPad2,5; iPhone 5 is iPhone 5,1. See http://theiphonewiki.com/wiki/index.php?title=Models
// OSX:                        returns "x86_64"
```

From the Cordova iOS related source code, I am expecting to see either a simulator's model identifier or the hardware machine which is one of the two items above. When I did a quick test of the source code on a macOS, `x86_64` is returned.

Basically, for an iOS app on a macOS, I am expecting to see the `iOS` platform on a `x86_64` model. I suspect on a Apple Silicon, it would say instead `iOS` platform on an `ARM64`.

I personally think this might be enough, but can anyone explain why it might not be?

**Comment by SailingSteve** [comment]
This is the object returned by device, when running on Apple Silicon.  (With the addition in this PR).

```
2020-10-19 09:28:00.803495-0700 We Vote[1530:17084] Dump Object window.device available: true
2020-10-19 09:28:00.803554-0700 We Vote[1530:17084] Dump Object window.device platform: iOS
2020-10-19 09:28:00.803605-0700 We Vote[1530:17084] Dump Object window.device version: 14.0
2020-10-19 09:28:00.803698-0700 We Vote[1530:17084] Dump Object window.device uuid: 4C9A03D3-4457-5253-8384-B4581A3F7396
2020-10-19 09:28:00.803778-0700 We Vote[1530:17084] Dump Object window.device cordova: 6.1.1
2020-10-19 09:28:00.804903-0700 We Vote[1530:17084] Dump Object window.device model: iPad8,9
2020-10-19 09:28:00.805041-0700 We Vote[1530:17084] Dump Object window.device manufacturer: Apple
2020-10-19 09:28:00.805150-0700 We Vote[1530:17084] Dump Object window.device isVirtual: false
2020-10-19 09:28:00.805631-0700 We Vote[1530:17084] Dump Object window.device serial: unknown
2020-10-19 09:28:00.806016-0700 We Vote[1530:17084] Dump Object window.device isiOSAppOnMac: true
```
Unfortunately 'iPad8,9' matches a real physical device, so it is not usable to detect isiOSAppOnMac condition.

Screen Shot 2020-10-19 at 9 36 37 AM

I need access to `isiOSAppOnMac` so that I can tell that the app covers the entire window, and that the hardware menu is not present (Wifi signal, Cell signal, battery, etc).

**Comment by SailingSteve** [comment]
@erisu I initially used cordova-plugins-diagnostic to find out that the architecture is ARM64.  Then I realized that both the phones and the Mac will be using ARM64 at some point, so the architecture can't be used to determine the equivalent of `isiOSAppOnMac`.

By the way, I have no attachment to the variable name `isiOSAppOnMac`, we can name it whatever we want, but I just defaulted to the name Apple chose.

File: www/device.js [hunk 1]
```diff
@@ -59,6 +60,7 @@ function Device () {
                 me.cordova = buildLabel;
                 me.model = info.model;
                 me.isVirtual = info.isVirtual;
+                me.isiOSAppOnMac = info.isiOSAppOnMac || false;
```
Comment (erisu) [inline www/device.js:63]:
```suggestion
                // isiOSAppOnMac is iOS specific. If defined, it will be appended.
                if (info.isiOSAppOnMac !== undefined) {
                    me.isiOSAppOnMac = info.isiOSAppOnMac;
                }
```

**Review by erisu — APPROVED** [review]
Sorry for the late reply.

I re-reviewed the changes and read your last response to my suggestion.

I agree with your statement that it may be possible that the phone's architecture will change to ARM64 in the future and that my workaround suggestion would eventually become invalid.

As for the variable name, it is acceptable.

I also posted a suggested change. Since this new property is iOS-specific, I don't know if we want it to be on the `window.device` object always.

For example, if the app was built only for Android, should the object display iOS-specific properties?

I think we could check if it is defined before appending it to the object.

Other then the change request, could you also rebase with the main branch and resolve the conflict.

Everything else appears good and can be merged in. Sorry again for the delay.

**Comment by SailingSteve** [comment]
@erisu I  adopted your suggestion, rebased, tested on a device simulator, and the desktop and it works fine.  The CI check is failing on a README.md merge problem, but the markdown looks good.  Maybe the ci needs to be forced to run again?

**Comment by breautek** [comment]
Closing/reopening PR to restart appveyor.

**Comment by breautek** [comment]
That didn't work..

Upon closer inspection, looks like your fork is out-dated from apace's master.

You're going to want to add apache's remote to your fork:

`git remote add apache https://github.com/apache/cordova-plugin-device.git`

Then fetch it:

`git fetch apache`

Then rebase apache's master into your master:

`git rebase apache/master`

You can do all of this on your own fork's master. During the rebase, you should hit the conflict that it's complaining about and solve it.

**Comment by erisu** [comment]
I should be able to merge this PR as soon as it is rebased and the conflicts are resolved.

**Comment by SailingSteve** [comment]
I'm sorry, I use git all day, but only in a simple and repetitive way.  I followed breauteks instructions, and they looked successful.  I had to fix the problem in the README.md.   But it now looks like my PR contains all the changes since I forked -- is that the way it should look?  

The good news, is that when I diff my branch with apache/master the changes look good -- ie the diff only shows the changes I want to have merged.

**Comment by breautek** [comment]
> I'm sorry, I use git all day, but only in a simple and repetitive way. 

No worries, I wasn't used to rebasing until I started contributing to Cordova/other open source projects myself. For the longest time I was afraid of rebasing (because it can be dangerous)

But the extra commits is still problematic. Not sure why the PR believes these are new commits. While the changes are okay, having all of these commits in this PR will be problematic if we ever have to revert this PR in the future for whatever reason.

It might be easier to create a new branch off of current master and cherry-pick the commits that is actually related to this PR.

I've gone ahead and created a [PR](https://github.com/SailingSteve/cordova-plugin-device/pull/1) against your own fork. You won't be able to merge it due to merge conflicts because you're fork's master is in a diverged state. But if the changes looks like they are matching what should be provided here, then there are instructions in the PR that should be enough to correct it the diverged state problem and once you force push your master, it should reflect here with your single commit encapsulating all your changes.

**Comment by breautek** [comment]
If you have any questions, feel free to ping me in [slack](https://join.slack.com/t/cordova/shared_invite/zt-z70vy6tx-7VNulesO0Qz0Od9QV4tc1Q)

**Comment by SailingSteve** [comment]
Thanks!  I'll do that on Thursday.
